cluster/loadbalance/p2c/loadbalance.go

```diff
@@ -89,49 +85,62 @@ func (l *p2cLoadBalance) Select(invokers []protocol.Invoker, invocation protocol
-	}
-
-	// TODO(justxuewei): It should have a strategy to drop some metrics after a period of time.
-	remainingJIface, err := m.GetMethodMetrics(invokers[j].GetURL(), methodName, metrics.HillClimbing)
-	if err != nil {
-		if errors.Is(err, metrics.ErrMetricsNotFound) {
-			logger.Debugf("[P2C select] The invoker[%d] was selected, because it hasn't been selected before.", j)
-			return invokers[j]
-		}
-		logger.Warnf("get method metrics err: %v", err)
-		return nil
-	}
-
-	// Convert interface to int, if the type is unexpected, panic immediately
-	remainingI, ok := remainingIIface.(uint64)
-	if !ok {
-		panic(fmt.Sprintf("[P2C select] the type of %s expects to be uint64, but gets %T",
-			metrics.HillClimbing, remainingIIface))
-	}
-
-	remainingJ, ok := remainingJIface.(uint64)
-	if !ok {
-		panic(fmt.Sprintf("the type of %s expects to be uint64, but gets %T", metrics.HillClimbing, remainingJIface))
-	}
 
-	logger.Debugf("[P2C select] The invoker[%d] remaining is %d, and the invoker[%d] is %d.", i, remainingI, j, remainingJ)
+	weightI, weightJ := Weight(invokers[i].GetURL(), invokers[j].GetURL(), methodName)
 
 	// For the remaining capacity, the bigger, the better.
-	if remainingI > remainingJ {
+	if weightI > weightJ {
 		logger.Debugf("[P2C select] The invoker[%d] was selected.", i)
 		return invokers[i]
 	}
 
 	logger.Debugf("[P2C select] The invoker[%d] was selected.", j)
 	return invokers[j]
 }
+
+//Weight w_i = s_i + ε*t_i
+func Weight(url1, url2 *common.URL, methodName string) (weight1, weight2 float64) {
+
+	s1 := successRateWeight(url1, methodName)
+	s2 := successRateWeight(url2, methodName)
+
+	rtt1Iface, _ := metrics.EMAMetrics.GetMethodMetrics(url1, methodName, metrics.RTT)
+	rtt2Iface, _ := metrics.EMAMetrics.GetMethodMetrics(url2, methodName, metrics.RTT)
+	rtt1 := metrics.ToFloat64(rtt1Iface)
+	rtt2 := metrics.ToFloat64(rtt2Iface)
+	logger.Debugf("[P2C Weight Metrics] [invoker1] %s's s score: %f, rtt: %f; [invoker2] %s's s score: %f, rtt: %f.",
+		url1.Ip, s1, rtt1, url2.Ip, s2, rtt2)
+	avgRtt := (rtt1 + rtt2) / 2
+	t1 := normalize((1 + avgRtt) / (1 + rtt1))
+	t2 := normalize((1 + avgRtt) / (1 + rtt2))
+
+	avgS := (s1 + s2) / 2
+	avgT := (t1 + t2) / 2
+	e := avgS / avgT
+
+	weight1 = s1 + e*t1
+	weight2 = s2 + e*t2
+	logger.Debugf("[P2C Weight] [invoker1] %s's s score: %f, t score: %f, weight: %f; [invoker2] %s's s score: %f, t score: %f, weight: %f.",
+		url1.Ip, s1, t1, weight1, url2.Ip, s2, t2, weight2)
+	return weight1, weight2
+}
+
+func normalize(x float64) float64 {
+	return 1.5 * x / (x + 1)
+}
+
+func successRateWeight(url *common.URL, methodName string) float64 {
+	requestsIface, _ := metrics.SlidingWindowCounterMetrics.GetMethodMetrics(url, methodName, metrics.Requests)
+	acceptsIface, _ := metrics.SlidingWindowCounterMetrics.GetMethodMetrics(url, methodName, metrics.Accepts)
+
+	requests := metrics.ToFloat64(requestsIface)
+	accepts := metrics.ToFloat64(acceptsIface)
+	r := (1 + accepts) / (1 + requests)
+
+	if r > 1 {
```

Comment:
What would make r greater than 1? IMO, if it does, something is going wrong. So an error output should be dumpped.